addon/PreviewProvider.js

```diff
@@ -208,38 +207,86 @@ PreviewProvider.prototype = {
+      let metadataLinkFavicons = null;
+      if (existingLinks.has(link.cache_key)) {
+        const cachedMetadataLink = existingLinks.get(link.cache_key);
+        metadataLinkFavicons = cachedMetadataLink.favicons;
+        enhancedLink.metadata_source = cachedMetadataLink.metadata_source;
         enhancedLink.title = cachedMetadataLink.title;
         enhancedLink.description = cachedMetadataLink.description;
         enhancedLink.provider_name = cachedMetadataLink.provider_name;
         enhancedLink.images = cachedMetadataLink.images;
-        enhancedLink.favicons = cachedMetadataLink.favicons;
-        enhancedLink.metadata_source = cachedMetadataLink.metadata_source;
-        if (cachedMetadataLink.favicons && cachedMetadataLink.favicons.length) {
-          enhancedLink.favicon_url = cachedMetadataLink.favicons[0].url;
-        }
       }
 
-      // Add tippy top data, if available
-      Object.assign(enhancedLink, this._tippyTopProvider.processSite(enhancedLink));
-
-      return enhancedLink;
-    }).filter(link => link);
+      // consolidate favicons, background color and metadata source, then return the final link
+      enhancedLink.favicon_url = this._consolidateFavicons(tippyTopFavicon, metadataLinkFavicons, firefoxFavicon);
+      enhancedLink.background_color = this._consolidateBackgroundColors(tippyTopBackgroundColor, firefoxBackgroundColor, metadataLinkFavicons);
+      results.push(enhancedLink);
+    }
 
     this._tabTracker.handlePerformanceEvent(event, "previewCacheHits", existingLinks.size);
     this._tabTracker.handlePerformanceEvent(event, "previewCacheMisses", processedLinks.length - existingLinks.size);
     return results;
   }),
 
+  /*
+   * Consolidate favicons from tippytop, firefox, and our own metadata. Tippytop
+   * is our number 1 choice, followed by the favicon returned by the metadata service,
+   * and finally firefox's data URI as a last resort
+   */
+  _consolidateFavicons(tippyTopFavicon, metadataFavicons, firefoxFavicon) {
+    const metadataFavicon = metadataFavicons && metadataFavicons[0] && metadataFavicons[0].url;
+    return tippyTopFavicon || metadataFavicon || firefoxFavicon;
+  },
+
+  /*
+   * Consolidate the background colors from tippytop, firefox, and our own metadata.
+   * TippyTop background color and metadata background color both need to be converted
+   * from hex to an rgb array, whereas the firefox background color is already rgb
+   */
+  _consolidateBackgroundColors(tippyTopBackgroundColor, firefoxBackgroundColor, metadataFavicons) {
+    const metadataBackgroundColor = metadataFavicons && metadataFavicons[0] && metadataFavicons[0].color;
+    return this._hexToRgb(tippyTopBackgroundColor) || firefoxBackgroundColor || this._hexToRgb(metadataBackgroundColor);
+  },
+
+  /**
+   * Convert a hex color string to the RGB form, e.g. #0A0102 => [10, 1, 2]
+   */
+  _hexToRgb(hex) {
+    let result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
```

Comment:
I know this is just copy/pasta, but we're always certain that the `#RRGGBB` will always be the full 7 characters and not the shortened `#RGB` syntax, right?